### PR TITLE
DM-1424 [App] Add Practice Adoption Locations (VAMCs) and Adoption Status

### PR DESCRIPTION
**Pull request by ahay-agile6**
### JIRA issue link
https://agile6.atlassian.net/browse/DM-1424

## Description - what does this code do?
Enables practice owners to be able to add Adoptions to their practice per facility.
They can add the date started and the date ended for each phase (except for In progress, only date started).

The fields for the form are:
Status
State
Facility
Date started (optional)
Date ended (optional)

## Testing done - how did you test it/steps on how can another person can test it 
pre-req: logged in as practice owner who can edit a practice
optional: map enabled via the google api key environment variable
1. Navigate to any practice that the user can edit
2. Click "Edit"
3. Make sure the "Adoptions" option is in the side nav, click it to navigate there
  1. if the practice does not have any Adoptions yet, the grey circle will be in the side nav
4. Click "Add new entry"
5. The practice create adoption form should appear
6. Fill out the required fields on the form
11. "In progress" status shouldn't let the user choose a "Date adoption finished"
11. Choosing "Completed" or "Unsuccessful" statuses should let the user choose a "Date adoption finished"
11. Click "Clear entry"
11. The form should be in the initial state
6. Fill out the required fields on the form
7. Click "Add entry"
8. If the practice didn't have any adoptions, the "Adoptions" in the sidenav should have a green check mark now!
9. update the entry with dates and such as you see fit
10. try different use cases like trying to make the start date later than the end date
10. there should be a client-side message that says as such
10. Click "Cancel edits"
10. The form should close and the "minus" icon should turn back into a "pencil and paper" icon in the accordion
10. Delete an entry by going into a created entry and clicking "Delete entry"
10. The page should refresh and a message saying the entry was delete should appear
10. Navigate to the practice show page to check the number of adoptions is correct

BONUS: Check the map with new adoptions created
1. Make sure the map feature is on via the environment variable "GOOGLE_API_KEY"
2. check the practice page
3. the new adoptions should appear on the map

## Screenshots, Gifs, Videos from application (if applicable)
localhost_3200_practices_4-sight_edit_adoptions

adoptions_spec_2

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
https://www.figma.com/file/DDUSxYiD1MaWNrGFmMjY8d8a/Diffusion-Marketplace-Designs?node-id=1745%3A27776

Prototype:
https://skylight.invisionapp.com/overview/Adoption-Editor-ck7z2timm0li6016slful4io4/screens?v=Cr9vTu1Qk5dwZ5JsJrTCQw%3D%3D&linkshare=urlcopied

## Acceptance criteria
check the story :) it's pretty long

## Definition of done
- [x] Unit tests written (if applicable)
- [x] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating JIRA issue
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/views/practices/_adoption_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+              end_year_el.classList.remove('dm-color-disabled');
+              end_year_el.disabled = false;
+            }
+          }
+
+          Array.prototype.forEach.call(radios, function(radio) {
+            radio.addEventListener('change', changeHandler);
+          });
+        <% end %>
+        <%= javascript_tag 'data-turbolinks-track': 'reload' do %>
```
Comment (emyl3) [inline app/views/practices/_adoption_form.html.erb:84]: do we need this `javascript_tag`?

File: app/views/practices/_show_authenticated.html.erb [hunk 2]
```diff
@@ -146,8 +146,9 @@
               <h4 class="margin-top-0 margin-bottom-1">Adoptions</h4>
               <%# <p><%= @practice.number_failed %> <%#unsuccessful</p> %>
               <p class="margin-0 font-sans-xs-2 line-height-sans-505">
-                <span><%= @practice.number_adopted.present? ? @practice.number_adopted : '0' %></span>
-                facilit<%= @practice.number_adopted == 1 ? 'y' : 'ies' %> have adopted this practice</p>
+                <% number_adopted = @practice.diffusion_histories.count %>
```
Comment (emyl3) [inline app/views/practices/_show_authenticated.html.erb:149]: ❓ So we are including unsuccessful adoptions as well?

File: app/views/practices/_adoption_form.html.erb (hunk 1 above)
Comment (ahay-agile6) [reply app/views/practices/_adoption_form.html.erb:84]: can probably combine, yeah

File: app/views/practices/_show_authenticated.html.erb (hunk 2 above)
Comment (ahay-agile6) [reply app/views/practices/_show_authenticated.html.erb:149]: gooooood question

File: app/views/practices/_adoption_form.html.erb (hunk 1 above)
Comment (ahay-agile6) [reply app/views/practices/_adoption_form.html.erb:84]: oh, or get rid of. must've missed it in my refactoring :smile:

**Comment by emyl3** [comment]
It looks like newly added completed adoptions are showing as "Completed" while existing adoptions are shown as "Complete" not sure if we want to be consistent here

Screen Shot 2020-04-08 at 1 58 17 PM

**Comment by ahay-agile6** [comment]
@emyl3 regretting not following my instincts and doing an `enum` for the statuses :laughing: 
gonna update that now~

**Comment by emyl3** [comment]
Can we clear the "Success" alert after clicking either of the two create adoption form buttons "Clear entry" and "Add entry" to avoid confusion?
Screen Shot 2020-04-08 at 2 56 39 PM

**Comment by ahay-agile6** [comment]
@emyl3 oh, yeah, sure!

File: app/assets/javascripts/practice_editor_adoptions.es6 [hunk 3]
```diff
@@ -27,6 +27,11 @@
         event.preventDefault();
         const formId = $(event.target).data('form-id');
         const formIdSelector = `#${formId}`;
+
+        // clear all alerts
```
Comment (emyl3) [inline app/assets/javascripts/practice_editor_adoptions.es6:31]: could we also clear the alert in the  `clearAdoptionEntryFrom` on click as well? Thank you!!

**Pull request merged**